File: notes/javanotes/src/test/java/com/example/notesapp/service/NoteServiceTest.java
package com.example.notesapp.service;

import com.example.notesapp.model.Note;
import com.example.notesapp.repository.InMemoryNoteRepository;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class NoteServiceTest {
    private final NoteService noteService = new NoteService(new InMemoryNoteRepository());

    @Test
    void createsNotesWithGeneratedIds() {
        Note note = noteService.createNote("Shopping", "Buy milk");

        assertEquals(1L, note.getId());
        assertEquals("Shopping", note.getTitle());
        assertEquals("Buy milk", note.getContent());
    }

    @Test
    void searchesByTitleAndContent() {
        noteService.createNote("Shopping", "Buy milk");
        noteService.createNote("Work", "Finish report");

        assertEquals(1, noteService.searchNotes("milk").size());
        assertEquals(1, noteService.searchNotes("work").size());
    }

    @Test
    void deletesExistingNotes() {
        Note note = noteService.createNote("Task", "Do it");

        assertTrue(noteService.deleteNote(note.getId()));
        assertFalse(noteService.deleteNote(note.getId()));
    }

    @Test
    void updatesExistingNotesWithoutChangingCreationTime() {
        Note note = noteService.createNote("Task", "Do it");
        Note updatedNote = noteService.updateNote(note.getId(), "Task updated", "Do it better");

        assertEquals(note.getId(), updatedNote.getId());
        assertEquals(note.getCreatedAt(), updatedNote.getCreatedAt());
        assertEquals("Task updated", updatedNote.getTitle());
        assertEquals("Do it better", updatedNote.getContent());
    }

    @Test
    void rejectsUnknownNotesWhenUpdating() {
        assertThrows(IllegalArgumentException.class, () -> noteService.updateNote(99L, "Title", "Body"));
    }

    @Test
    void rejectsBlankTitles() {
        assertThrows(IllegalArgumentException.class, () -> noteService.createNote(" ", "Body"));
    }
}

File: notes/flutternotes/test/note_service_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:notes_app/repositories/in_memory_note_repository.dart';
import 'package:notes_app/services/note_service.dart';

void main() {
  late NoteService noteService;

  setUp(() {
    noteService = NoteService(InMemoryNoteRepository());
  });

  test('creates notes with generated ids', () async {
    final note = await noteService.createNote('Shopping', 'Buy milk');

    expect(note.id, 1);
    expect(note.title, 'Shopping');
    expect(note.content, 'Buy milk');
  });

  test('searches by title and content', () async {
    await noteService.createNote('Shopping', 'Buy milk');
    await noteService.createNote('Work', 'Finish report');

    expect((await noteService.searchNotes('milk')).length, 1);
    expect((await noteService.searchNotes('work')).length, 1);
  });

  test('deletes existing notes', () async {
    final note = await noteService.createNote('Task', 'Do it');

    expect(await noteService.deleteNote(note.id), isTrue);
    expect(await noteService.deleteNote(note.id), isFalse);
  });

  test('updates existing notes without changing creation time', () async {
    final note = await noteService.createNote('Task', 'Do it');
    final updated = await noteService.updateNote(note.id, 'Task updated', 'Do it better');

    expect(updated.id, note.id);
    expect(updated.createdAt, note.createdAt);
    expect(updated.title, 'Task updated');
    expect(updated.content, 'Do it better');
  });

  test('rejects unknown notes when updating', () async {
    expect(
      () => noteService.updateNote(99, 'Title', 'Body'),
      throwsArgumentError,
    );
  });

  test('rejects blank titles', () async {
    expect(
      () => noteService.createNote(' ', 'Body'),
      throwsArgumentError,
    );
  });
}
